### PR TITLE
Extract capture config from grok patterns up front

### PR DESCRIPTION
**Pull request by nik9000**
This extracts the configuration for extracting values from a groked
string when building the grok expression to do two things:
1. Create a method exposing that configuration on `Grok` itself which
   will be used grok `grok` flavored runtime fields.
2. Marginally speed up extracting grok values by skipping a little
   string manipulation.

### CODE REVIEW

File: libs/grok/src/main/java/org/elasticsearch/grok/Grok.java [hunk 1]
```diff
@@ -151,7 +158,7 @@ private void forbidCircularReferences(String patternName, List<String> path, Str
         }
     }
 
-    public String groupMatch(String name, Region region, String pattern) {
+    private String groupMatch(String name, Region region, String pattern) {
```
Comment (nik9000) [inline libs/grok/src/main/java/org/elasticsearch/grok/Grok.java:161]: This didn't need to be public and it confused me that it was.

File: libs/grok/src/main/java/org/elasticsearch/grok/Grok.java [hunk 2]
```diff
@@ -167,7 +174,7 @@ public String groupMatch(String name, Region region, String pattern) {
      *
      * @return named regex expression
      */
-    public String toRegex(String grokPattern) {
+    protected String toRegex(String grokPattern) {
```
Comment (nik9000) [inline libs/grok/src/main/java/org/elasticsearch/grok/Grok.java:177]: This *feels* private, but it is used by tests. I tightened the visibility while I was there, but not didn't refactor a bunch of tests.

File: libs/grok/src/main/java/org/elasticsearch/grok/GrokMatchGroup.java [hunk 3]
```diff
@@ -1,68 +0,0 @@
...
- * software distributed under the License is distributed on an
- * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
- * KIND, either express or implied.  See the License for the
- * specific language governing permissions and limitations
- * under the License.
- */
-
-package org.elasticsearch.grok;
-
-final class GrokMatchGroup {
```
Comment (nik9000) [inline libs/grok/src/main/java/org/elasticsearch/grok/GrokMatchGroup.java:22]: Moved this into the two new classes.

**Review by jakelandis — APPROVED** [review]
LGTM

**Review by benwtrent — APPROVED** [review]
the separation leaves it much more readable. 

:shipit:

**Comment by nik9000** [comment]
Thanks @benwtrent and @jakelandis !

**Pull request merged**